### PR TITLE
Differentiate socks5h from socks5 and socks4a from socks4

### LINKED ISSUE #1035: Differentiate socks5h from socks5 and socks4a from socks4 when handling proxy string
In a proxy string, socks5h:// and socks4a:// mean that the hostname is
resolved by the SOCKS server.  socks5:// and socks4:// mean that the
hostname is resolved locally.  socks4a:// means to use SOCKS4a, which is
an extension of SOCKS4.  Let's make urllib3 honor it.

### PR DESCRIPTION
**Pull request by ijcoe6ru**
In the proxy string, socks5h:// and socks4a:// mean that the hostname is
to be resolved by the socks server.  socks5:// and socks4:// mean the
hostname is to be resolved locally.

Fix #1035

### CODE REVIEW

File: test/contrib/test_socks.py [hunk 1]
```diff
@@ -229,6 +229,37 @@ def request_handler(listener):
...
         self.assertEqual(response.headers['Server'], 'SocksTestServer')
 
+    def test_local_dns(self):
+        def request_handler(listener):
+            sock = listener.accept()[0]
+
+            handler = handle_socks5_negotiation(sock, negotiate=False)
+            addr, port = next(handler)
+
+            self.assertIn(addr, ['127.0.0.1', '::1'])
```
Comment (Lukasa) [inline test/contrib/test_socks.py]: You will need to start using `unittest2` in these test cases to use `assertIn`: it's not present on Python 2.6.

**Comment by ijcoe6ru** [comment]
I've added tests for doing DNS lookup locally and changed the tests that
use example.com to use socks4a and socks5h.

As of the concern that some code may break, they'll probably be fine
since some well-known software, such as curl, use this convention.

File: test/contrib/test_socks.py (hunk 1 above)
Comment (ijcoe6ru) [inline test/contrib/test_socks.py]: Excuse me.  Why are the tests run in Python 2.6 instead of 2.7?

Comment (SethMichaelLarson) [inline test/contrib/test_socks.py]: We have to support Python 2.6 because our upstreams require it.

Comment (Lukasa) [inline test/contrib/test_socks.py]: We don't just run in Python 2.6. We run in Python 2.6, Python 2.7, Python 3.3, Python 3.4, Python 3.5, and PyPy 2, across two different operating systems. =) The tests, and the code, must work on all of those platforms.

File: test/contrib/test_socks.py [hunk 2]
```diff
@@ -236,7 +236,12 @@ def request_handler(listener):
             handler = handle_socks5_negotiation(sock, negotiate=False)
             addr, port = next(handler)
 
-            self.assertIn(addr, ['127.0.0.1', '::1'])
+            # We're not using assertIn because the tests need to pass under the
+            # old Python 2.6
+            if not addr in ['127.0.0.1', '::1']:
+                raise self.failureException(
+                    "Requested address is %s, 127.0.0.1 or ::1 expected" %
+                    safe_repr(addr))
```
Comment (Lukasa) [inline test/contrib/test_socks.py]: There's no need to do this. You can leave the `self.assertIn` statement and then just edit `dummyserver/testcase.py` to change `import unittest` to

```python
import sys

if sys.version_info >= (2, 7):
    import unittest
else:
    import unittest2 as unittest
```

That should resolve the problem.

**Comment by ijcoe6ru** [comment]
Thanks.  I hadn't noticed that unittest2 was used elsewhere, so I thought it would've introduced new dependencies.

File: dummyserver/testcase.py [hunk 3]
```diff
@@ -1,4 +1,8 @@
-import unittest
+import sys
+if sys.version_info >= (2, 7):
+    import unittest
+else:
+    import unittest2 as unittest
```
Comment (Lukasa) [inline dummyserver/testcase.py]: So to make flake8 happy, let's move this block down to line 19.

**Comment by ijcoe6ru** [comment]
I don't know why the test has failed. Can anyone help me?

**Comment by Lukasa** [comment]
It's just a transient failure, I've restarted that test.

**Comment by SethMichaelLarson** [comment]
To see details about which tests failed and on what platform, you can click on the Travis details in the required checks box here at the bottom of the PR.  It looks like the test that failed was a randomly failure.

**Comment by salehe** [comment]
This is a very useful feature. Looking forward for it to be merged

**Comment by ijcoe6ru** [comment]
Why hasn't this been merged yet?

**Comment by SethMichaelLarson** [comment]
Huzzah! :D

**Pull request merged**